src/backend/distributed/utils/metadata_cache.c

```diff
@@ -194,13 +195,68 @@ IsDistributedTable(Oid relationId)
 		return false;
 	}
 
-	cacheEntry = LookupDistTableCacheEntry(relationId);
+	/*
+	 * We will not use cache if there is a version mismatch, because while
+	 * creating the cache, we assume our metadata tables are in particular
+	 * structure. However, in version mismatch, this is not guaranteed, thus
+	 * we will read from pg_dist_partition table and we will only assume
+	 * specific details which are most likely do not change in long term.
+	 */
+	installedVersion = InstalledExtensionVersion();
+	if (!MajorVersionsCompatible(installedVersion, CITUS_EXTENSIONVERSION))
```

Comment:
InstalledExtensionVersion already caches the result and we set it to NULL in case of [CREATE/ALTER EXTENSION](https://github.com/citusdata/citus/blob/master/src/backend/distributed/utils/metadata_cache.c#L1072)